airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -226,6 +226,10 @@ def _read_incremental(
         )
         logger.debug(f"Processing stream slices for {stream_name}", extra={"stream_slices": slices})
         total_records_counter = 0
+        if not slices:
+            # Safety net to ensure we always emit at least one state message even if there are no slices
+            checkpoint = self._checkpoint_state(stream_instance, stream_instance.state, connector_state)
```

Comment:
@girarda, there is an on call issue that seems related to this new change.
https://sentry.io/organizations/airbytehq/issues/3521477693/?project=6527718
https://github.com/airbytehq/oncall/issues/466

> 'Events' object has no attribute 'state'

I don't understand the CDK very well. But it looks like the `stream_instance.state` is only set conditionally:

```py
if stream_state and "state" in dir(stream_instance):
  stream_instance.state = stream_state
```

Should the `state` be set for more cases?

Reply:
@tuliren alex is on PTO - I've asked someone from the team to assist

Reply:
Got it. Thank you.